src/circuit/gadget/ecc/chip/mul/incomplete.rs

```diff
@@ -0,0 +1,296 @@
+        &self,
+        region: &mut Region<'_, pallas::Base>,
+        offset: usize,
+        base: &EccPoint,
+        bits: &[Option<bool>],
+        acc: (X<pallas::Base>, Y<pallas::Base>, Z<pallas::Base>),
+    ) -> Result<(X<pallas::Base>, Y<pallas::Base>, Vec<Z<pallas::Base>>), Error> {
+        // Check that we have the correct number of bits for this double-and-add.
+        assert_eq!(bits.len(), self.num_bits);
+
+        // Handle exceptional cases
+        let (x_p, y_p) = (base.x.value(), base.y.value());
+        let (x_a, y_a) = (acc.0.value(), acc.1 .0);
+
+        if let (Some(x_a), Some(y_a), Some(x_p), Some(y_p)) = (x_a, y_a, x_p, y_p) {
+            // A is point at infinity
+            if (x_p == pallas::Base::zero() && y_p == pallas::Base::zero())
+            // Q is point at infinity
+            || (x_a == pallas::Base::zero() && y_a == pallas::Base::zero())
+            // x_p = x_a
+            || (x_p == x_a)
+            {
+                return Err(Error::SynthesisError);
+            }
+        }
+
+        // Enable `q_mul` on all but the last row of the incomplete range.
+        for row in 1..(self.num_bits - 1) {
+            self.q_mul.enable(region, offset + row)?;
+        }
+
+        // Initialise the running `z` sum for the scalar bits.
+        let mut z = copy(region, || "starting z", self.z, offset, &acc.2, &self.perm)?;
+
+        // Increase offset by 1; we used row 0 for initializing `z`.
+        let offset = offset + 1;
+
+        // Define `x_p`, `y_p`
+        let x_p = base.x.value();
+        let y_p = base.y.value();
+
+        // Initialise acc
+        let mut x_a = copy(
+            region,
+            || "starting x_a",
+            self.x_a,
+            offset,
+            &acc.0,
+            &self.perm,
+        )?;
+        let mut y_a = *acc.1;
+
+        // Initialise vector to store all interstitial `z` running sum values.
+        let mut zs: Vec<Z<pallas::Base>> = Vec::new();
+
+        // Incomplete addition
+        for (row, k) in bits.iter().enumerate() {
+            // z_{i} = 2 * z_{i+1} + k_i
+            let z_val = z.value().zip(k.as_ref()).map(|(z_val, k)| {
+                pallas::Base::from_u64(2) * z_val + pallas::Base::from_u64(*k as u64)
+            });
+            let z_cell = region.assign_advice(
+                || "z",
+                self.z,
+                row + offset,
+                || z_val.ok_or(Error::SynthesisError),
+            )?;
+            z = CellValue::new(z_cell, z_val);
+            zs.push(Z(z));
+
+            // Assign `x_p`, `y_p`
+            region.assign_advice(
+                || "x_p",
+                self.x_p,
+                row + offset,
+                || x_p.ok_or(Error::SynthesisError),
+            )?;
+            region.assign_advice(
+                || "y_p",
+                self.y_p,
+                row + offset,
+                || y_p.ok_or(Error::SynthesisError),
+            )?;
```

Comment:
This causes a double-assignment in the circuit, due to sharing these columns between `hi_config` and `lo_config`. This is fine, as we pass the same `base` into both `incomplete_addition` invocations.